publish-mqtt/src/main.rs

```diff
@@ -241,35 +170,25 @@ async fn requests(requests_tx: Sender<Request>, device_base: &str) -> Result<(),
                 let mac_address = device.get_address()?;
                 let name = sensor_names.get(&mac_address).unwrap_or(&mac_address);
                 println!(
-                    "{} ({}) Temperature: {:.2}ºC Humidity: {:?}% Battery {} mV ({} %)",
+                    "{} Temperature: {:.2}ºC Humidity: {:?}% Battery {} mV ({} %) ({})",
                     device.get_id(),
-                    name,
                     temperature,
                     humidity,
                     battery_voltage,
-                    battery_percent
+                    battery_percent,
+                    name
                 );
 
                 let node_id = mac_address.replace(":", "");
-                let node_base = format!("{}/{}", device_base, node_id);
-                publish_retained(
-                    &requests_tx,
-                    format!("{}/temperature", node_base),
-                    &format!("{:.2}", temperature),
-                )
-                .await?;
-                publish_retained(
-                    &requests_tx,
-                    format!("{}/humidity", node_base),
-                    &humidity.to_string(),
-                )
-                .await?;
-                publish_retained(
-                    &requests_tx,
-                    format!("{}/battery", node_base),
-                    &battery_percent.to_string(),
-                )
-                .await?;
+                homie
+                    .publish_value(&node_id, "temperature", &format!("{:.2}", temperature))
+                    .await?;
+                homie
+                    .publish_value(&node_id, "humidity", &humidity.to_string())
+                    .await?;
+                homie
+                    .publish_value(&node_id, "battery", &battery_percent.to_string())
```

Comment:
This is a bit ugly. Publish::new() is defined as:
```
impl Publish {
    pub fn new<S: Into<String>, P: Into<Vec<u8>>>(topic: S, qos: QoS, payload: P) -> Publish {
        Publish {
            dup: false,
            qos,
            retain: false,
            pkid: 0,
            topic: topic.into(),
            payload: Bytes::from(payload.into()),
        }
    }
...
}
```

Could we make our thing accept `impl Into<String>` or something? Do we care if the float temperature has some bullshit trailing bits?

Reply:
erm...
```
    value: impl ToString,
```

Reply:
I think this is how I would do it.

```
diff --git a/publish-mqtt/src/homie.rs b/publish-mqtt/src/homie.rs
index 1049843..86ede0b 100644
--- a/publish-mqtt/src/homie.rs
+++ b/publish-mqtt/src/homie.rs
@@ -16,7 +16,7 @@ pub struct Property {
 impl Property {
     pub fn new(id: &str, name: &str, datatype: &str, unit: &str) -> Property {
         Property {
-            id: id.to_string(),
+            id: id.to_owned(),
             name: name.to_string(),
             datatype: datatype.to_string(),
             unit: unit.to_string(),
@@ -182,7 +182,7 @@ impl HomieDevice {
         &self,
         node_id: &str,
         property_id: &str,
-        value: &str,
+        value: impl ToString,
     ) -> Result<(), SendError<Request>> {
         publish_retained(
             &self.requests_tx,
@@ -196,8 +196,9 @@ impl HomieDevice {
 async fn publish_retained(
     requests_tx: &Sender<Request>,
     name: String,
-    value: &str,
+    value: impl ToString,
 ) -> Result<(), SendError<Request>> {
+    let value: String = value.to_string();
     let mut publish = Publish::new(name, QoS::AtLeastOnce, value);
     publish.set_retain(true);
     requests_tx.send(publish.into()).await
diff --git a/publish-mqtt/src/main.rs b/publish-mqtt/src/main.rs
index 211b94c..3e5f0e4 100644
--- a/publish-mqtt/src/main.rs
+++ b/publish-mqtt/src/main.rs
@@ -181,13 +181,11 @@ async fn requests(mut homie: HomieDevice) -> Result<(), Box<dyn Error>> {
 
                 let node_id = mac_address.replace(":", "");
                 homie
-                    .publish_value(&node_id, "temperature", &format!("{:.2}", temperature))
+                    .publish_value(&node_id, "temperature", format!("{:.2}", temperature))
                     .await?;
+                homie.publish_value(&node_id, "humidity", humidity).await?;
                 homie
-                    .publish_value(&node_id, "humidity", &humidity.to_string())
-                    .await?;
-                homie
-                    .publish_value(&node_id, "battery", &battery_percent.to_string())
+                    .publish_value(&node_id, "battery", battery_percent)
                     .await?;
             } else {
                 println!("Invalid value from {}", device.get_id());
```

or if I want to minimise the number of generic functions, I might write it like:
```
diff --git a/publish-mqtt/src/homie.rs b/publish-mqtt/src/homie.rs
index 1049843..c13945f 100644
--- a/publish-mqtt/src/homie.rs
+++ b/publish-mqtt/src/homie.rs
@@ -16,7 +16,7 @@ pub struct Property {
 impl Property {
     pub fn new(id: &str, name: &str, datatype: &str, unit: &str) -> Property {
         Property {
-            id: id.to_string(),
+            id: id.to_owned(),
             name: name.to_string(),
             datatype: datatype.to_string(),
             unit: unit.to_string(),
@@ -182,12 +182,12 @@ impl HomieDevice {
         &self,
         node_id: &str,
         property_id: &str,
-        value: &str,
+        value: impl ToString,
     ) -> Result<(), SendError<Request>> {
         publish_retained(
             &self.requests_tx,
             format!("{}/{}/{}", self.device_base, node_id, property_id),
-            value,
+            &value.to_string(),
         )
         .await
     }
diff --git a/publish-mqtt/src/main.rs b/publish-mqtt/src/main.rs
index 211b94c..3e5f0e4 100644
--- a/publish-mqtt/src/main.rs
+++ b/publish-mqtt/src/main.rs
@@ -181,13 +181,11 @@ async fn requests(mut homie: HomieDevice) -> Result<(), Box<dyn Error>> {
 
                 let node_id = mac_address.replace(":", "");
                 homie
-                    .publish_value(&node_id, "temperature", &format!("{:.2}", temperature))
+                    .publish_value(&node_id, "temperature", format!("{:.2}", temperature))
                     .await?;
+                homie.publish_value(&node_id, "humidity", humidity).await?;
                 homie
-                    .publish_value(&node_id, "humidity", &humidity.to_string())
-                    .await?;
-                homie
-                    .publish_value(&node_id, "battery", &battery_percent.to_string())
+                    .publish_value(&node_id, "battery", battery_percent)
                     .await?;
             } else {
                 println!("Invalid value from {}", device.get_id());

```

Reply:
Done the second option.